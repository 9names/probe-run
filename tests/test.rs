use os_pipe::pipe;
use std::{
    io::Read,
    process::{Command, ExitStatus},
    sync::Mutex,
};
use structopt::lazy_static::lazy_static;

struct RunResult {
    exit_status: ExitStatus,
    output: String,
}

lazy_static! {
    /// rust will try to run the tests in parallel by default, and `insta` doesn't like the
    /// usual way of disabling this via `--test-threads=1`, so we're using this
    /// mutex to make sure we're not re-flashing until the last run is finished
    static ref ONE_RUN_AT_A_TIME: Mutex<i32> = Mutex::new(0i32);
}

/// run probe-run with `args` and truncate the "Finished .. in .." and "Running `...`" flashing output
/// NOTE: this currently only capures `stdin`, so any `log::` ed output, like flashing
fn run_and_truncate(args: &str) -> RunResult {
    let _guard = ONE_RUN_AT_A_TIME.lock().unwrap();

    // add prefix to run this repository's version of `probe-run` and
    // remove user-dependent registry and rustc information from backtrace paths
    let complete_command = format!("run -- {} --shorten-paths", args);

    let (mut reader, writer) = pipe().unwrap();
    let writer_clone = writer.try_clone().unwrap();

    let mut handle = Command::new("cargo")
        .args(complete_command.split(" "))
        // capture stderr and stdout while preserving line order
        .stdout(writer)
        .stderr(writer_clone)
        // run `probe-run`
        .spawn()
        .unwrap();

    // retrieve output and clean up
    let mut probe_run_output = String::new();
    reader.read_to_string(&mut probe_run_output).unwrap();
    let exit_status = handle.wait().unwrap();

    // remove the lines printed during flashing, as they contain timing info that's not always the same
    let output = probe_run_output
        .lines()
        .filter(|line| {
            !line.starts_with("    Finished")
            && !line.starts_with("     Running `")
            && !line.starts_with("    Blocking waiting for file lock ")
            && !line.starts_with("   Compiling probe-run v")
            // TODO don't drop the `└─ probe_run @ ...` locations after
            // https://github.com/knurling-rs/probe-run/issues/217 is resolved
            && !line.starts_with("└─ ")
        })
        .map(|line| format!("{}\n", line))
        .collect();

    RunResult {
        exit_status,
        output,
    }
}

#[test]
// this test should not be run by default, as it requires the target hardware to be present
#[ignore]
fn successful_run_has_no_backtrace() {
    let run_result = run_and_truncate("--chip nRF52840_xxAA tests/test_elfs/hello");

    assert_eq!(true, run_result.exit_status.success());
    insta::assert_snapshot!(run_result.output);
}


#[test]
// this test should not be run by default, as it requires the target hardware to be present
#[ignore]
fn successful_run_can_enforce_backtrace() {
    let run_result =
        run_and_truncate("--chip nRF52840_xxAA tests/test_elfs/hello --backtrace=always");

    assert_eq!(true, run_result.exit_status.success());
    insta::assert_snapshot!(run_result.output);
}

#[test]
// this test should not be run by default, as it requires the target hardware to be present
#[ignore]
fn stack_overflow_is_reported_as_such() {
    let run_result = run_and_truncate("--chip nRF52840_xxAA tests/test_elfs/overflow");

    assert_eq!(false, run_result.exit_status.success());
    insta::assert_snapshot!(run_result.output);
}

#[test]
// this test should not be run by default, as it requires the target hardware to be present
#[ignore]
fn panic_is_reported_as_such() {
    let run_result = run_and_truncate("--chip nRF52840_xxAA tests/test_elfs/panic");

    assert_eq!(false, run_result.exit_status.success());
    insta::assert_snapshot!(run_result.output);
}

#[test]
// this test should not be run by default, as it requires the target hardware to be present
#[ignore]
fn panic_verbose() {
    // record current verbose backtrace to catch deviations
    let run_result = run_and_truncate("--chip nRF52840_xxAA tests/test_elfs/panic --verbose");

    assert_eq!(false, run_result.exit_status.success());
    insta::assert_snapshot!(run_result.output);
}

#[test]
// this test should not be run by default, as it requires the target hardware to be present
#[ignore]
fn unsuccessful_run_can_suppress_backtrace() {
    let run_result =
        run_and_truncate("--chip nRF52840_xxAA tests/test_elfs/panic --backtrace=never");

    assert_eq!(false, run_result.exit_status.success());
    insta::assert_snapshot!(run_result.output);
}

#[test]
// this test should not be run by default, as it requires the target hardware to be present
#[ignore]
fn stack_overflow_can_suppress_backtrace() {
    let run_result = run_and_truncate("--chip nRF52840_xxAA tests/test_elfs/overflow --backtrace=never");

    assert_eq!(false, run_result.exit_status.success());
    insta::assert_snapshot!(run_result.output);
}