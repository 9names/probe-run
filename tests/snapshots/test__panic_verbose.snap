---
source: tests/test.rs
expression: run_result.output

---
(HOST) DEBUG vector table: VectorTable { location: 0, initial_stack_pointer: 2003fbc8, reset: 101, hard_fault: 13e5 }
(HOST) DEBUG RAM region: 0x20000000-0x2003FFFF
(HOST) DEBUG section `.data` is in RAM at 0x2003FBC8 ..= 0x2003FBF7
(HOST) DEBUG section `.bss` is in RAM at 0x2003FBF8 ..= 0x2003FBFB
(HOST) DEBUG section `.uninit` is in RAM at 0x2003FBFC ..= 0x2003FFFB
(HOST) DEBUG valid SP range: 0x20000000 ..= 0x2003FBC7
(HOST) DEBUG found 1 probes
(HOST) DEBUG opened probe
(HOST) DEBUG started session
(HOST) INFO  flashing program (5.82 KiB)
(HOST) INFO  success!
(HOST) DEBUG 261062 bytes of stack available (0x20000000 ..= 0x2003FBC7), using 1024 byte canary
(HOST) TRACE setting up canary took 0.013s (75.91 KiB/s)
(HOST) DEBUG starting device
(HOST) DEBUG Successfully attached RTT
────────────────────────────────────────────────────────────────────────────────
 INFO  main
 ERROR panicked at 'explicit panic'
────────────────────────────────────────────────────────────────────────────────
(HOST) TRACE reading canary took 0.012s (83.93 KiB/s)
(HOST) DEBUG stack canary intact
(HOST) DEBUG LR=0xFFFFFFF9 PC=0x000013E4
(HOST) DEBUG LR=0x000006FF PC=0x0000071C
(HOST) DEBUG update_cfa: CFA changed Some(2003fb68) -> 2003fb70
(HOST) DEBUG LR=0x00000709 PC=0x000006FE
(HOST) DEBUG update_cfa: CFA changed Some(2003fb70) -> 2003fb78
(HOST) DEBUG LR=0x000005C5 PC=0x00000708
(HOST) DEBUG update_cfa: CFA changed Some(2003fb78) -> 2003fb80
(HOST) DEBUG LR=0x00000229 PC=0x000005C4
(HOST) DEBUG update_cfa: CFA changed Some(2003fb80) -> 2003fbb0
(HOST) DEBUG LR=0x00000195 PC=0x00000228
(HOST) DEBUG update_cfa: CFA changed Some(2003fbb0) -> 2003fbb8
(HOST) DEBUG LR=0x00000233 PC=0x00000194
(HOST) DEBUG update_cfa: CFA changed Some(2003fbb8) -> 2003fbc0
(HOST) DEBUG LR=0x0000018B PC=0x00000232
(HOST) DEBUG update_cfa: CFA changed Some(2003fbc0) -> 2003fbc8
(HOST) DEBUG LR=0xFFFFFFFF PC=0x0000018A
stack backtrace:
   0: HardFaultTrampoline
      <exception entry>
   1: lib::inline::__udf
        at ./asm/inline.rs:172:5
   2: __udf
        at ./asm/lib.rs:49:17
   3: cortex_m::asm::udf
        at [cortex-m-0.7.2]/src/asm.rs:43:5
   4: _defmt_panic
        at /Users/lottesteenbrink/ferrous/my-app/src/lib.rs:13:5
   5: defmt::export::panic
        at [defmt-0.2.2]/src/export.rs:233:14
   6: panic::__cortex_m_rt_main
        at /Users/lottesteenbrink/ferrous/my-app/src/bin/panic.rs:10:5
   7: main
        at /Users/lottesteenbrink/ferrous/my-app/src/bin/panic.rs:6:1
   8: ResetTrampoline
        at [cortex-m-rt-0.6.14]/src/lib.rs:547:26
   9: Reset
        at [cortex-m-rt-0.6.14]/src/lib.rs:550:13
(HOST) ERROR the program panicked

